subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyModifier.java

```diff
@@ -88,18 +96,37 @@ default Provider<? extends MinimalExternalModuleDependency> modify(ProviderConve
      * @param providerToDependency the provider
      * @return a provider to the modified dependency
      */
-    default <D extends ModuleDependency> Provider<D> modify(Provider<D> providerToDependency) {
+    public final <D extends ModuleDependency> Provider<D> modify(Provider<D> providerToDependency) {
         return providerToDependency.map(this::modify);
     }
 
     /**
      * Takes a given {@link ModuleDependency} and modifies the dependency to select the variant of the given module. Dependency resolution may fail if the given module does not have a compatible variant.
-     * <p><br></p>
+     *
+     * <p>
+     * The dependency will be copied, so the original dependency will not be modified.
+     * </p>
      *
      * @param dependency the dependency to modify
-     * @return the modified dependency
      * @param <D> the type of the {@link ModuleDependency}
+     * @return the modified dependency
+     */
+    public final <D extends ModuleDependency> D modify(D dependency) {
+        // Enforce a copy of the dependency to avoid modifying the original dependency
+        // The unchecked cast can be incorrect if D is an implementation type, but it shouldn't be used in that way.
+        @SuppressWarnings("unchecked")
+        D copy = (D) dependency.copy();
+        modifyImpl(copy);
+        return copy;
+    }
+
+    /**
+     * Modify the given dependency.
+     *
+     * @param dependency the dependency to modify
      * @implSpec This method must be implemented.
+     * @since 8.4
      */
-    <D extends ModuleDependency> D modify(D dependency);
+    protected abstract void modifyImpl(ModuleDependency dependency);
```

Comment:
❌ We need to update the class-level docs here to explain how to implement a modifier since you no longer override the `modify(D)` and you don't need to avoid implementing the default methods.

💭 I would also consider a slightly different name:
- `doModify` - this is similar to what we use internally
- `modifyImplementation` - this describes this as the "implementation of modify(...)"
- `modifyDependency` - this is what it does